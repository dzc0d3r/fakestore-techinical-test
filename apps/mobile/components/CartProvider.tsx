// components/CartProvider.tsx
import AsyncStorage from "@react-native-async-storage/async-storage";
import React, { createContext, useContext, useEffect, useState } from "react";

export interface CartItem {
  id: number;
  name: string;
  price: number;
  quantity: number;
  image: string;
}

interface CartContextValue {
  cartItems: CartItem[];
  addToCart: (item: CartItem) => void;
  removeFromCart: (item: CartItem) => void;
  clearCart: () => void;
  getCartTotal: () => number;
  deleteFromCart: (itemId: number) => void;
}

const CartContext = createContext<CartContextValue>({} as CartContextValue);

export const CartProvider = ({ children }: { children: React.ReactNode }) => {
  const [cartItems, setCartItems] = useState<CartItem[]>([]);

  useEffect(() => {
    const loadCart = async () => {
      try {
        const storedCart = await AsyncStorage.getItem("cart");
        if (storedCart) setCartItems(JSON.parse(storedCart));
      } catch (error) {
        console.error("Error loading cart:", error);
      }
    };
    loadCart();
  }, []);

  const saveCart = async (items: CartItem[]) => {
    try {
      await AsyncStorage.setItem("cart", JSON.stringify(items));
    } catch (error) {
      console.error("Error saving cart:", error);
    }
  };

  const addToCart = (item: CartItem) => {
    const existingItem = cartItems.find((i) => i.id === item.id);
    let newItems;

    if (existingItem) {
      newItems = cartItems.map((i) =>
        i.id === item.id ? { ...i, quantity: i.quantity + 1 } : i,
      );
    } else {
      newItems = [...cartItems, { ...item, quantity: 1 }];
    }

    setCartItems(newItems);
    saveCart(newItems);
  };

  const removeFromCart = (item: CartItem) => {
    const existingItem = cartItems.find((i) => i.id === item.id);
    let newItems;

    if (existingItem?.quantity === 1) {
      newItems = cartItems.filter((i) => i.id !== item.id);
    } else {
      newItems = cartItems.map((i) =>
        i.id === item.id ? { ...i, quantity: i.quantity - 1 } : i,
      );
    }

    setCartItems(newItems);
    saveCart(newItems);
  };

  const clearCart = () => {
    setCartItems([]);
    saveCart([]);
  };

  const deleteFromCart = (itemId: number) => {
    const newItems = cartItems.filter((i) => i.id !== itemId);
    setCartItems(newItems);
    saveCart(newItems);
  };

  const getCartTotal = () => {
    return cartItems.reduce(
      (total, item) => total + item.price * item.quantity,
      0,
    );
  };

  return (
    <CartContext.Provider
      value={{
        cartItems,
        addToCart,
        removeFromCart,
        clearCart,
        getCartTotal,
        deleteFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => useContext(CartContext);
