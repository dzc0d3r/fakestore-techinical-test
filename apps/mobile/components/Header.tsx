import { useAuth } from "@/hooks/useAuth";
import { Ionicons } from "@expo/vector-icons";
import React from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";

interface HeaderProps {
  title: string;
  onMenuPress: () => void;
}

const Header: React.FC<HeaderProps> = ({ title, onMenuPress }) => {
  const { token, logout } = useAuth();

  return (
    <SafeAreaView style={styles.safeArea}>
      <View style={styles.headerContainer}>
        <TouchableOpacity onPress={onMenuPress} style={styles.iconButton}>
          <Ionicons name="menu" size={28} color="#000" />
        </TouchableOpacity>

        {token ? (
          <TouchableOpacity onPress={logout} style={styles.iconButton}>
            <Ionicons name="log-out-outline" size={28} color="#000" />
          </TouchableOpacity>
        ) : (
          // Placeholder for alignment if not logged in
          <View style={styles.iconPlaceholder} />
        )}
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  safeArea: {
    backgroundColor: "#fff",
  },
  headerContainer: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    paddingHorizontal: 16,
    paddingVertical: 12,
    borderBottomWidth: 1,
    borderBottomColor: "#e0e0e0",
  },
  iconButton: {
    padding: 4,
  },
  title: {
    fontSize: 18,
    fontWeight: "bold",
    color: "#000",
  },
  iconPlaceholder: {
    width: 32,
  },
});

export default Header;
