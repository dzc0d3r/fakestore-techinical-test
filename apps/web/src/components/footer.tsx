import { auth } from "@/auth";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import Link from "next/link";


export async function Footer() {
  const session = await auth();
  return (
    <footer className="bg-gray-800 py-12 text-white">
      <div className="container mx-auto px-4">
        <div className="grid grid-cols-1 gap-8 md:grid-cols-4">
          <div className="space-y-4">
            <h3 className="text-xl font-bold">Company</h3>
            <ul className="space-y-2">
              <li>
                <Link href="/about" className="hover:text-gray-300">
                  About Us
                </Link>
              </li>
              <li>
                <Link href="/privacy" className="hover:text-gray-300">
                  Privacy Policy
                </Link>
              </li>
            </ul>
          </div>

          <div className="space-y-4">
            <h3 className="text-xl font-bold">Shop</h3>
            <ul className="space-y-2">
              <li>
                <Link href="/products" className="hover:text-gray-300">
                  All Products
                </Link>
              </li>
              <li>
                {session ? (
                  <Link href="/profile" className="hover:text-gray-300">
                    Profile
                  </Link>
                ) : (
                  <Link href="/login" className="hover:text-gray-300">
                    My Profile
                  </Link>
                )}
              </li>
            </ul>
          </div>

          <div className="space-y-4">
            <h3 className="text-xl font-bold">Contact</h3>
            <ul className="space-y-2">
              <li>Email: support@example.com</li>
              <li>Phone: (123) 456-7890</li>
            </ul>
          </div>

          <div className="space-y-4">
            <h3 className="text-xl font-bold">Newsletter</h3>
            <p className="text-gray-300">
              Subscribe to our newsletter for updates
            </p>
            <div className="flex gap-2">
              <Input
                placeholder="Your email"
                className="border-gray-700 bg-gray-800"
              />
              <Button variant="secondary">Subscribe</Button>
            </div>
          </div>
        </div>

        <div className="mt-8 border-t border-gray-800 pt-8 text-center text-gray-400">
          <p>Â© {new Date().getFullYear()} Weasydoo. All rights reserved.</p>
        </div>
      </div>
    </footer>
  );
}
