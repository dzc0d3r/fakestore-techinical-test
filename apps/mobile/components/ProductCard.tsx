import { ThemedText } from "@/components/ThemedText";
import { ThemedView } from "@/components/ThemedView";
import type { Product } from "api";
import { Image, StyleSheet, TouchableOpacity } from "react-native";

export const ProductCard = ({ product }: { product: Product }) => {
  return (
    <ThemedView style={styles.card}>
      <Image
        source={{ uri: product.image }}
        style={styles.image}
        resizeMode="contain"
      />
      <ThemedView style={styles.details}>
        <ThemedText type="defaultSemiBold">{product.name}</ThemedText>
        <ThemedText style={styles.price}>${product.price}</ThemedText>
        <ThemedText numberOfLines={2} style={styles.description}>
          {product.description}
        </ThemedText>
      </ThemedView>
    </ThemedView>
  );
};

const styles = StyleSheet.create({
  card: {
    flexDirection: "row",
    borderRadius: 8,
    padding: 12,
    gap: 12,
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
    elevation: 2,
  },
  image: {
    width: 100,
    height: 100,
    borderRadius: 4,
  },
  details: {
    flex: 1,
    gap: 4,
  },
  price: {
    fontSize: 16,
    fontWeight: "bold",
    color: "#0a7ea4",
  },
  description: {
    fontSize: 12,
    opacity: 0.8,
  },
});
