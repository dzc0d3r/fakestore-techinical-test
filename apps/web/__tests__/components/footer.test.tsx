import { Footer } from "@/components/footer";
import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";


// Mock authentication functions
vi.mock("@/auth", () => ({
  auth: vi.fn().mockResolvedValue({ user: { role: "user" } }),
  signOut: vi.fn(),
}));

describe("Footer", () => {
  it("should match snapshot", async () => {
    // Await the Footer component to get the resolved React element.
    const footerElement = await Footer();
    const { container } = render(footerElement);
    expect(container).toMatchSnapshot();
  });
});