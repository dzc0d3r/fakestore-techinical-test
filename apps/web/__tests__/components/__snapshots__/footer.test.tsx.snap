// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should match snapshot 1`] = `
<div>
  <footer
    class="bg-gray-800 py-12 text-white"
  >
    <div
      class="container mx-auto px-4"
    >
      <div
        class="grid grid-cols-1 gap-8 md:grid-cols-4"
      >
        <div
          class="space-y-4"
        >
          <h3
            class="text-xl font-bold"
          >
            Company
          </h3>
          <ul
            class="space-y-2"
          >
            <li>
              <a
                class="hover:text-gray-300"
                href="/about"
              >
                About Us
              </a>
            </li>
            <li>
              <a
                class="hover:text-gray-300"
                href="/privacy"
              >
                Privacy Policy
              </a>
            </li>
          </ul>
        </div>
        <div
          class="space-y-4"
        >
          <h3
            class="text-xl font-bold"
          >
            Shop
          </h3>
          <ul
            class="space-y-2"
          >
            <li>
              <a
                class="hover:text-gray-300"
                href="/products"
              >
                All Products
              </a>
            </li>
            <li>
              <a
                class="hover:text-gray-300"
                href="/profile"
              >
                My Account
              </a>
            </li>
          </ul>
        </div>
        <div
          class="space-y-4"
        >
          <h3
            class="text-xl font-bold"
          >
            Contact
          </h3>
          <ul
            class="space-y-2"
          >
            <li>
              Email: support@example.com
            </li>
            <li>
              Phone: (123) 456-7890
            </li>
          </ul>
        </div>
        <div
          class="space-y-4"
        >
          <h3
            class="text-xl font-bold"
          >
            Newsletter
          </h3>
          <p
            class="text-gray-300"
          >
            Subscribe to our newsletter for updates
          </p>
          <div
            class="flex gap-2"
          >
            <input
              class="file:text-foreground placeholder:text-muted-foreground focus-visible:ring-ring flex h-9 w-full rounded-md border px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium focus-visible:outline-none focus-visible:ring-1 disabled:cursor-not-allowed disabled:opacity-50 md:text-sm border-gray-700 bg-gray-800"
              placeholder="Your email"
            />
            <button
              class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-9 px-4 py-2"
            >
              Subscribe
            </button>
          </div>
        </div>
      </div>
      <div
        class="mt-8 border-t border-gray-800 pt-8 text-center text-gray-400"
      >
        <p>
          Â© 
          2025
           Weasydoo. All rights reserved.
        </p>
      </div>
    </div>
  </footer>
</div>
`;
